update.check/src/main/java/org/kaazing/gateway/update/check/UpdateCheckGatewayObserver.java

```diff
@@ -0,0 +1,111 @@
+import org.kaazing.gateway.util.InternalSystemProperty;
+import org.kaazing.gateway.util.scheduler.SchedulerProvider;
+
+/**
+ * Creates and manages periodic checks to see if the gateway has updates
+ */
+public class UpdateCheckGatewayObserver extends GatewayObserverFactorySpiPrototype {
+
+    private static final String ENTERPRISE_URL = "https://version.kaazing.com";
+    private static final String COMMUNITY_URL = "https://version.kaazing.org";
+    private GatewayVersion currentVersion;
+    private String productName;
+    private String versionServiceUrl;
+    private Set<UpdateCheckListener> listeners = new HashSet<>();
+    private GatewayVersion latestVersion;
+
+    public UpdateCheckGatewayObserver() {
+        addListener(new UpdateCheckLoggingListener());
+    }
+
+    /**
+     * @return latest @GatewayVersion or null if the latest version has not been discovered
+     */
+    private synchronized GatewayVersion getLatestGatewayVersion() {
+        return latestVersion;
+    }
+
+    protected void setLatestGatewayVersion(GatewayVersion newlatestVersion) {
+        if (latestVersion != null && latestVersion.compareTo(newlatestVersion) >= 0) {
+            return;
+        }
+
+        latestVersion = newlatestVersion;
+        if (newlatestVersion.compareTo(currentVersion) <= 0) {
+            return;
+        }
+        for (UpdateCheckListener listener : listeners) {
+            listener.newVersionAvailable(currentVersion, getLatestGatewayVersion());
+        }
+    }
+
+    /**
+     * Adds an @UpdateCheckListener who will be notified when the version changes,
+     * @param newListener
+     */
+    public void addListener(UpdateCheckListener newListener) {
+        if (newListener == null) {
+            throw new IllegalArgumentException("newListener");
+        }
+        GatewayVersion latestGatewayVersion = getLatestGatewayVersion();
+        if (latestGatewayVersion != null && latestGatewayVersion.compareTo(currentVersion) > 0) {
+            newListener.newVersionAvailable(currentVersion, latestGatewayVersion);
+        }
+        listeners.add(newListener);
+    }
+
+    @Override public void startingGateway(GatewayContext gatewayContext) {
+        Map<String, Object> injectables = gatewayContext.getInjectables();
+        Properties properties = (Properties) injectables.get("configuration");
+        if (!InternalSystemProperty.UPDATE_CHECK.getBooleanProperty(properties)) {
+            return;
+        }
+
+        productName = getGatewayProductTitle().replaceAll("\\s+", "");
+        currentVersion = parseGatewayVersion(getGatewayProductVersionPatch());
+
+        String serviceUrl = InternalSystemProperty.SERVICE_URL.getProperty(properties);
+        if (serviceUrl != null) {
+            versionServiceUrl = serviceUrl;
+        } else {
+            versionServiceUrl =
+                    (getGatewayProductEdition().toLowerCase().contains("enterprise")) ? ENTERPRISE_URL : COMMUNITY_URL;
```

Comment:
nit.: The outer parentheses are extraneous. ?: operator has the lowest precedence, so it will execute last.